---
title: "Mutational Signature Comprehensive Analysis Toolkit"
author:
- name: Aaron Chevalier
  affiliation: &id Boston University School of Medicine
- name: Joshua Campbell
  affiliation: *id
  email: camp@bu.edu
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Mutational Signature Comprehensive Analysis Toolkit}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE, results = "asis"}
BiocStyle::markdown()
knitr::opts_chunk$set(echo = TRUE, dev = "png")
```

# Introduction
A variety of exogenous exposures or endogenous biological processes can contribute to the overall mutational load observed in human tumors. Many different mutational patterns, or “mutational signatures”, have been identified across different tumor types. These signatures can provide a record of environmental exposure and can give clues about the etiology of carcinogenesis. The Mutational Signature TookKit has utilities for extracting variants from a variety of file formats, contains multiple methods for discovery of novel signatures or prediction of known signatures, as well as many types of downstream visualizations for exploratory analysis. This package has the ability to parse and combine multiple motif classes in the mutational signature discovery or prediction processes. Mutation motifs include single base substitutions (SBS), double base substitutions (DBS), insertions (INS) and deletions (DEL).

# Installation
Currently musicatk can be installed from on Bioconductor using the following code:

```{r, eval= FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE)){
    install.packages("BiocManager")}
BiocManager::install("musicatk")
```

To install the latest version from Github, use the following code:

```{r, eval = FALSE}
library(devtools)
install_github("campbio/musicatk")
```

The package can be loaded using the `library` command.

```{r, eval = TRUE, message = FALSE}
library(musicatk)
```

# Setting up a musica object
In order to discover or predict mutational signatures, we must first set up
our musica object by 1) extracting variants from files or objects such as
VCFs and MAFs, 2) selecting the appropriate reference genome 3) creating a musica object, and 4) building a count tables for our varinats of interest.

## Extracting variants
Variants can be extracted from various formats using the following functions:

* The `extract_variants_from_vcf_file()` function will extract variants from a [VCF](https://samtools.github.io/hts-specs/) file. The file will be imported using the readVcf function from the [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) package and then the variant information will be extracted from this object.
* The `extract_variants_from_vcf()` function extracts variants from a `CollapsedVCF` or `ExpandedVCF` object from the [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) package.
* The `extract_variants_from_maf_file()` function will extract variants from a file in [Mutation Annotation Format (MAF)](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/) used by TCGA.
* The `extract_variants_from_maf()` function will extract variants from a MAF object created by the [maftools](https://www.bioconductor.org/packages/release/bioc/html/maftools.html) package. 
* The `extract_variants_from_matrix()` function will get the information from a matrix or data.frame like object that has columns for the chromosome, start position, end position, reference allele, and mutation allele. 
* The `extract_variants()` function will extract variants from a list of objects. These objects can be any combination of VCF files, VariantAnnotation objects, MAF files, MAF objects, and data.frame objects.

Below are some examples of extracting variants from MAF and VCF files:

```{r extract_variants, message = FALSE}
# Extract variants from a MAF File
lusc_maf <- system.file("testdata", "public_TCGA.LUSC.maf", package = "musicatk") 
lusc.variants <- extract_variants_from_maf_file(maf_file = lusc_maf)

# Extract variants from an individual VCF file
luad_vcf <- system.file("testdata", "public_LUAD_TCGA-97-7938.vcf", 
                         package = "musicatk")
luad.variants <- extract_variants_from_vcf_file(vcf_file = luad_vcf)

# Extract variants from multiple files and/or objects
melanoma_vcfs <- list.files(system.file("testdata", package = "musicatk"), 
                           pattern = glob2rx("*SKCM*vcf"), full.names = TRUE)
variants <- extract_variants(c(lusc_maf, luad_vcf, melanoma_vcfs))
```

## Choosing a genome
musicatk uses [BSgenome](https://bioconductor.org/packages/release/bioc/html/BSgenome.html) objects to access genome sequence information that flanks each mutation which is used bases for generating mutation count tables. BSgenome objects store full genome sequences for different organisms. A full list of supported organisms can be obtained by running ```available.genomes()```. Custom genomes can be forged as well (see [BSgenome](https://bioconductor.org/packages/release/bioc/html/BSgenome.html) documentation). musicatk provides a utility function called ```select_genome()``` to allow users to quickly select human genome build versions "hg19" and "hg38" or mouse genome builds "mm9" and "mm10". The reference sequencing for these genomes are in UCSC format (e.g. chr1).

```{r select_genome}
g <- select_genome("hg38")
```

## Creating a musica object
The last preprocessing step is to create an object with the variants and the genome. This function will perform checks to ensure that the chromosome names and reference alleles in the input variant object match those in supplied BSgenome object. These checks can be turned off by setting `check_ref_chromosomes = FALSE` and `check_ref_bases = FALSE`, respectively.

```{r create_musica}
musica <- create_musica(x = variants, genome = g)
```

# Creating mutation count tables
Motifs are the building blocks of mutational signatures. Motifs themselves are 
a mutation combined with other genomic information. For instance, **SBS96** 
motifs are constructed from an SBS mutation and one upsteam and one downstream 
base sandwiched together. We build tables by counting these motifs for each 
sample. 
```{r build_tables}
build_standard_table(musica, g = g, table_name = "SBS96")
```

Here is a list of mutation tables that can be created:

* SBS96 - description
* table2 - description
* table3 - description

# Discover Signatures/Exposures
Discovery and prediction result are loaded into a self-contained result object 
that includes signatures and sample exposures.
```{r discover_sigs}
result <- discover_signatures(musica = musica, table_name = "SBS96", 
                              num_signatures = 3, method = "lda", nstart = 10)
```

# Plotting

## Signatures
```{r, plot_sigs}
plot_signatures(result)
```

Signatures can be named based on prior knowledge and displayed in the plots:

```{r, name_sigs}
name_signatures(result, c("Smoking", "APOBEC", "UV"))
plot_signatures(result)
```

## Exposures
**Exposures** are defined as the estimated amount of each signature in each tumor. Barplots showing the exposures in each sample can be plotted:
```{r exposures_raw}
plot_exposures(result, proportional = FALSE)
```

The proportion of each exposure in each tumor can be shown by setting ```proprotional = TRUE```:
```{r exposures_prop}
plot_exposures(result, proportional = TRUE)
```

Counts for individual samples can also be plotted:

```{r sample_counts}
samples <- get_sample_names(musica)
plot_sample_counts(musica, "SBS96", samples[1])
```

## Comparison to external signatures (e.g. COSMIC)
Compare to COSMIC signatures by leaving the second result as default

```{r compare_cosmic}
compare_cosmic_v2(result, threshold = 0.72)
```

# Predicting exposures using existing signatures
We can list which signatures are present in each tumor type according to the [COSMIC V2 database](https://cancer.sanger.ac.uk/cosmic/signatures_v2.tt). Here we find which signatures are present in lung cancer:
```{r subtype_map}
cosmic_v2_subtype_map("lung")
```

Exposures for a subset of COSMIC signatures 4, 11, 12, 15 in our current dataset:

```{r predict_cosmic}
pred_cosmic <- predict_exposure(musica = musica, table_name = "SBS96",
                               signature_res = cosmic_v2_sigs,
                               signatures_to_use =  c(12, 4, 11, 15),
                               algorithm = "lda")

plot_signatures(pred_cosmic)
plot_exposures(pred_cosmic, proportional = FALSE)
```

Note that if `signatures_to_use` is not supplied by the user, then exposures for all signatures in the result object will be estimated. We can predict exposures from any result object. For example, we can predict using the result object containing the discovered signatures:

```{r predict_previous}
pred_our_sigs <- predict_exposure(musica = musica, table_name = "SBS96",
                                 signature_res = result, algorithm = "lda")

```

Different signatures in different result objects can also be compared to each other:

```{r predict_compare}
compare_results(result = pred_cosmic, other_result = pred_our_sigs, threshold = 0.60)
```

# Use of sample annotations for advanced sample comparisons
## Adding annotations
```{r annotations}
sample_annotations <- read.table(system.file("testdata", 
                                             "sample_annotations.txt", 
                                             package = "musicatk"), sep = "\t", 
                                 header=TRUE)
init_sample_annotations(musica)
add_sample_annotations(musica = musica, annotations = sample_annotations, 
                       sample_column = "Sample_Names", 
                       columns_to_add = "Tumor_Subtypes")
```

## Standard discovery using musicatk with sample annotations
```{r plot_annotations}
#Add tumor type annotations to our samples
res <- discover_signatures(musica, table_name = "SBS96", num_signatures = 3)

plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes")
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", 
                             proportional = FALSE)
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", 
                             by_group = FALSE)
```


# Note on reproducibility

Several functions make use of stochastic algorithms or procedures which 
require the use of random number generator (RNG) for simulation or sampling. 
To maintain reproducibility, all these functions use a **default seed of 1** to 
make sure same results are generated each time one of these functions is 
called. Explicitly setting the `seed` arguments is needed for greater control 
and randomness. These functions include *discover_signatures*, 
*predict_exposure*, and *create_umap*.


# Session Information
```{r session}
sessionInfo()
```
