<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analysis of mutational signatures with the Shiny graphical user interface (GUI) • musicatk</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of mutational signatures with the Shiny graphical user interface (GUI)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">musicatk</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="https://www.camplab.net/musicatk" aria-label="Home"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/articles/installation.html">Installation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/articles/tutorial_tcga_cnsl.html">Analysis in the R console</a></li>
    <li><a class="dropdown-item" href="../articles/articles/tutorial_tcga_ui.html">Analysis with the user interface</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/campbio/musicatk" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/camplab1" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Analysis of mutational signatures with the Shiny graphical user interface (GUI)</h1>
                        <h4 data-toc-skip class="author">Nathan
Sahelijo, Zainab Khurshid, Tong Tong, Aaron Chevalier, Natasha Gurevich,
Joshua Campbell</h4>
            
            <h4 data-toc-skip class="date">Compiled January 14,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/campbio/musicatk/blob/HEAD/vignettes/articles/tutorial_tcga_ui.Rmd" class="external-link"><code>vignettes/articles/tutorial_tcga_ui.Rmd</code></a></small>
      <div class="d-none name"><code>tutorial_tcga_ui.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>A variety of exogenous exposures or endogenous biological processes
can contribute to the overall mutational load observed in human tumors.
Many different mutational patterns, or “mutational signatures”, have
been identified across different tumor types. These signatures can
provide a record of environmental exposure and can give clues about the
etiology of carcinogenesis. The Mutational Signature Comprehensive
Analysis Toolkit (musicatk) contains a complete end-to-end workflow for
characterization of mutational signatures in a cohort of samples.
musicatk has utilities for extracting variants from a variety of file
formats, multiple methods for discovery of novel signatures or
prediction of pre-existing signatures, and many types of downstream
visualizations for exploratory analysis. This package has the ability to
parse and combine multiple motif classes in the mutational signature
discovery or prediction processes. Mutation motifs include single base
substitutions (SBS), double base substitutions (DBS), insertions (INS)
and deletions (DEL). To start the app on your local machine, open
RStudio or R in the terminal and type the following in the console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/musicatk/">musicatk</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/musicatk.html">musicatk</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="importing-mutational-data">Importing mutational data<a class="anchor" aria-label="anchor" href="#importing-mutational-data"></a>
</h2>
<p>Users can upload their datasets through 1 of 3 different methods. -
“Import variants from files” - upload MAF and VCF files containing the
variant data. - “Import variants from TCGA” - easily import all open
access TCGA tumor datasets. - “Import previously generated musica
objects” - upload existing musica objects. <br></p>
<div class="section level3">
<h3 id="import-variants-from-files">Import variants from files<a class="anchor" aria-label="anchor" href="#import-variants-from-files"></a>
</h3>
<p>In order to discover or predict mutational signatures, we must first
set up our musica object by 1) extracting variants from files or objects
such as VCFs and MAFs, 2) selecting the appropriate reference genome 3)
creating a musica object, 4) adding sample-level annotations, and 5)
building a count tables for our variants of interest. Variants stored in
VCF or MAF files can be selected from your computation using the
‘browse’ button under the ‘Select Files’ label.</p>
<p><img src="articles/ui_screenshots/import.jpg" alt="screenshot of import page" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Clicking the ‘Add Samples’ button will add the files to the list of
samples to be imported. The delete and undo button next to the file can
be pressed to remove it or revert the change respectively. After adding
the files, the ‘Import’ button can be clicked to start the importing
process.</p>
<p><img src="articles/ui_screenshots/import1.jpg" alt="screenshot of added files" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Once the process is completed, a notification in the bottom right
will appear. A variants table from the uploaded files will also appear
which can be downloaded by pressing the ‘Download Variants’ button.</p>
<p><img src="articles/ui_screenshots/import2.jpg" alt="screenshot of variants table from uploaded files" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="import-tcga-datasets">Import TCGA Datasets<a class="anchor" aria-label="anchor" href="#import-tcga-datasets"></a>
</h3>
<p>This tab allows for selection of single or multiple datasets
generated in The Cancer Genome Atlas (TCGA). The original TCGA page for
the tumor name list can be referred to by selecting the ‘Full Tumor
List’ link on the top. Once the desired datasets have been selected, the
‘Import’ button at the bottom can be clicked to download the TCGA data.
It may take some time to download and process all of the selected
datasets. Once finished, a completion notification will appear at the
bottom right.</p>
<p><img src="articles/ui_screenshots/TCGA.jpg" alt="screenshot of TCGA import page" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="import-previously-generated-musica-objects">Import previously generated musica objects<a class="anchor" aria-label="anchor" href="#import-previously-generated-musica-objects"></a>
</h3>
<p>This tab can be used to import a user’s previously generated musica
variant object in .rda or .rds format. The browse button can be selected
for selecting files. Once selected, a bar saying file upload is
completed will appear. Once finished, a notification message in the
bottom right will appear along with a table of variants. If a musica
variant object was uploaded then the users can skip the Create Musica
tab and move to the build tables tab directly.</p>
<p><img src="articles/ui_screenshots/upload_musica.jpg" alt="screenshot of import musica page" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="creating-a-musica-object">Creating a musica object<a class="anchor" aria-label="anchor" href="#creating-a-musica-object"></a>
</h2>
<p>In this tab users can process variants and create a musica variant
object which can be used for discovery of signatures or prediction of
exposures. The reference genome can be selected from the drop down menu
labeled “Choose Genome” that matches the genome build of the variants
and the ‘Create Musica Object’ button can then be clicked. Once it’s
done, a completion notification message at the bottom right and a
variant summary table will appear. The musica object can be downloaded
by clicking the ‘Download Musica Object’ button and re-uploaded to save
time in subsequent sessions.</p>
<p><img src="articles/ui_screenshots/create_musica.jpg" alt="screenshot of create musica object page" style="border: 1px solid rgba(0,0,0,0.2)"><br><img src="articles/ui_screenshots/create_musica1.jpg" alt="screenshot of variant summary table" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level2">
<h2 id="importing-sample-annotations">Importing sample annotations<a class="anchor" aria-label="anchor" href="#importing-sample-annotations"></a>
</h2>
<p>Sample annotations can be used to store information about each sample
such as tumor type or treatment status. They are optional and can be
used in downstream plotting functions such as <a href="../../docs/reference/plot_exposures.html"><em>plot_exposures</em></a>
or <a href="../../docs/reference/plot_umap.html"><em>plot_umap</em></a>
to group or color samples by a particular annotation.</p>
<p><img src="articles/ui_screenshots/AnnotationsTab.jpg" alt="screenshot of adding annotations" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Annotations are sample-level variables such as Tumor Type or Age.
These can be provided in a character delimited text file which must
include a column containing the sample IDs.</p>
<p><img src="articles/ui_screenshots/annotation_file.jpg" alt="screenshot of example annotations" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The sample IDs must match the sample names in the musica object. NA
values will be given to any samples not present in the annotation
file.</p>
<p><img src="articles/ui_screenshots/AnnotationOptions.jpg" alt="screenshot of annotation file upload complete" style="border: 1px solid rgba(0,0,0,0.2)"><br> After selecting the correct delimiter, a data table will appear
below to show the annotations that will be added to the musica object.
If your annotation file does not contain a header, unselect the “Header”
radio button. Choose the column that contains the sample names from the
“Sample Name Columns” dropdown and then click “Add Annotation”.</p>
<p><img src="articles/ui_screenshots/AnnotationTable.jpg" alt="screenshot of data table containing annotations" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level2">
<h2 id="creating-mutation-count-tables">Creating mutation count tables<a class="anchor" aria-label="anchor" href="#creating-mutation-count-tables"></a>
</h2>
<div class="section level3">
<h3 id="create-standard-tables">Create standard tables<a class="anchor" aria-label="anchor" href="#create-standard-tables"></a>
</h3>
<p>The “Build Tables” tab generates count tables for different mutation
schemas. These schemas are the input to the mutational signature
discovery and prediction functions. To build the count tables, the user
must select 1 of the 5 standard motifs in the “Select Count Table”
dropdown.</p>
<p><br></p>
<p>SBS96 - Motifs are the six possible single base pair mutation types
times the four possibilities each for upstream and downstream context
bases (464 = 96 motifs) <br> SBS192_Trans - Motifs are an extension of
SBS96 multiplied by the transcriptional strand
(translated/untranslated), can be specified with “Transcript_Strand”.
<br> SBS192_Rep - Motifs are an extension of SBS96 multiplied by the
replication strand (leading/lagging), can be specified with
“Replication_Strand”. <br> DBS - Motifs are the 78 possible
double-base-pair substitutions. <br> INDEL - Motifs are 83 categories
intended to capture different categories of indels based on base-pair
change, repeats, or microhomology, insertion or deletion, and
length.</p>
<p><br></p>
<p>In addition to selecting a motif, the user must also provide the
reference genome.</p>
<p><img src="articles/ui_screenshots/TableTab.jpg" alt="screenshot of build tables tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="combine-tables">Combine tables<a class="anchor" aria-label="anchor" href="#combine-tables"></a>
</h3>
<p>Any two or more tables can be combined into a new table. The option
to build a combined table will only appear after the second table has
been created.</p>
<p>&lt;img src=“vignettes/ui_screenshots/combine-table.alt =”screenshot
of combining tables” jpg” style=“border: 1px solid
rgba(0,0,0,0.2)”&gt; <br></p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-signatures-and-exposures">Estimating signatures and exposures<a class="anchor" aria-label="anchor" href="#estimating-signatures-and-exposures"></a>
</h2>
<p>The user has the option of discovering novel signatures and
exposures, or using previously discovered signatures to predict
exposures. To discover signatures use the “Discover” tab. To predict
signature exposures, use the “Predict” tab.</p>
<div class="section level3">
<h3 id="discovery-of-signatures-and-exposures">Discovery of signatures and exposures<a class="anchor" aria-label="anchor" href="#discovery-of-signatures-and-exposures"></a>
</h3>
<p>Mutational signatures and exposures are discovered using a Latent
Dirichlet Allocation (LDA) or a Non-Negative Matrix Factorization (NMF)
algorithms. These algorithms will deconvolute the mutation count matrix
into two matrices: 1) a “signature” matrix containing the probability of
each mutation type in each sample and 2) an “exposure” matrix containing
the estimated counts for each signature in each sample. Select a
modality, algorithm, number of signatures, and specify the name of the
result list and model to store the results and then click “Discover
signatures”.</p>
<p><img src="articles/ui_screenshots/DiscoverTab.jpg" alt="screenshot of discover tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="predicting-exposures-from-existing-signatures">Predicting exposures from existing signatures<a class="anchor" aria-label="anchor" href="#predicting-exposures-from-existing-signatures"></a>
</h3>
<p>Exposures for samples can be predicted using an existing set of
signatures stored in a result model object. Algorithms available for
prediction include “lda” and “decompTumor2Sig”.</p>
<p><img src="articles/ui_screenshots/PredictTab.jpg" alt="screenshot of predict tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The “Signatures to Predict” dropdown contains all the signatures in
the result object selected from the “Result to Predict” dropdown. You
can search this dropdown and select multiple signatures. In our example
we use the “lda” algorithm to predict exposures of 6 Cosmic
signatures.</p>
<p><img src="articles/ui_screenshots/PredictDropdown.jpg" alt="screenshot of signatures to predict options" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-results">Visualization of results<a class="anchor" aria-label="anchor" href="#visualization-of-results"></a>
</h2>
<p>The data visualization tab can be used to generate customized plots
for signatures and exposures. Plots will be made interactive if the
“plotly” option is selected.</p>
<div class="section level3">
<h3 id="plot-signatures">Plot signatures<a class="anchor" aria-label="anchor" href="#plot-signatures"></a>
</h3>
<p>Bar plots can be used to display the probability of each type of
mutation within each signature.</p>
<p><img src="articles/ui_screenshots/visual_sig.jpg" alt="screenshot of signature section of data visualization tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Previously, 8 single-base signatures were identified from a cohort of
lung adenocarcinoma, lung squamous cell carcinoma, and skin cutaneous
melanoma samples from TCGA. By default, signatures are named by numbers,
but an option of renaming signatures is provided if you want to rename
them according to their possible etiology (e.g. Smoking) or closest
correlation to COSMIC (e.g. SBS4).</p>
<p><img src="articles/ui_screenshots/sig_plot.jpg" alt="screenshot of plotted signatures" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="plot-exposures">Plot exposures<a class="anchor" aria-label="anchor" href="#plot-exposures"></a>
</h3>
<p>Bar plots, box plots, violin plots and scatter plots can be used to
visualize the exposure levels of each signature in each sample.</p>
<p><img src="articles/ui_screenshots/visual_exp.jpg" alt="screenshot of exposure section of data visualization tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>By default, a stacked bar plot sorted by the total number of
mutations is used. Each stacked bar shows the proportion of exposure of
each signature.</p>
<p><img src="articles/ui_screenshots/exp_plot1.jpg" alt="screenshot of plotted exposures" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The stacked bar plot can be ordered by signatures. If Signatures is
selected in the Sort By option, a bucket list will show up to allow you
select the signatures you want to use by dragging them from the left box
to the right box. Users can also set a limit on the number of samples to
display.</p>
<p><img src="articles/ui_screenshots/sorting.jpg" alt="screenshot of sorting options" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This stacked bar plot is now ordered by the exposure of signature 1
and only the top 400 samples were included here.</p>
<p><img src="articles/ui_screenshots/exp_plot2.jpg" alt="screenshot of sorted predicted exposures" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Box plot and violin plot can be used to visualize the distribution of
exposures or compare exposures between different groups of samples.</p>
<p><img src="articles/ui_screenshots/exp_box_setting1.jpg" alt="screenshot of settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>By default, a box plot of exposure for each signature will be
shown.</p>
<p><img src="articles/ui_screenshots/exp_plot3.jpg" alt="screenshot of boxplot of exposures" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If an annotation file is provided, then you can visually compare the
exposure of signatures among different groups.</p>
<p><img src="articles/ui_screenshots/exp_box_setting2.jpg" alt="screenshot of settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>In this box plot, exposures of each signature were grouped by tumor
types. We can find that signatures 1 and 5 had high activities in lung
cancer samples, while signatures 3 and 8 had high activities in skin
cancer samples.</p>
<p><img src="articles/ui_screenshots/exp_plot4.jpg" alt="screenshot of exposures boxplot, grouped by tumor type" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>We can also group samples by signatures and then color by tumor
types.</p>
<p><img src="articles/ui_screenshots/exp_box_setting3.jpg" alt="screenshot of settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This plot can be used to visualize and compare the distribution of
exposure levels for signature across an annotation such as tumor
type.</p>
<p><img src="articles/ui_screenshots/exp_plot5.jpg" alt="screenshot of exposures boxplot, grouped by signature" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="downstream-analyses">Downstream analyses<a class="anchor" aria-label="anchor" href="#downstream-analyses"></a>
</h2>
<div class="section level3">
<h3 id="compare-signatures">Compare Signatures<a class="anchor" aria-label="anchor" href="#compare-signatures"></a>
</h3>
<p>A common analysis is to compare discovered signatures estimated in a
dataset to those generated in other datasets or to those in the <a href="https://cancer.sanger.ac.uk/cosmic/signatures" class="external-link">COSMIC
database</a>. The threshold acts as a cutoff similarity score and can be
any value between 0 and 1. Results will populate in a downloadable data
table.</p>
<p><img src="articles/ui_screenshots/CompareTab.jpg" alt="screenshot of compare signatures tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Using the cosine metric with a threshold of 0.8,7 pairs of similar
signatures from COSMIC. Our Signature1 had a high cosine similarity to
Cosmic SBS4 indicating that some of our tumors have been exposed to
cigarette smoke.</p>
<p><img src="articles/ui_screenshots/CompareTable.jpg" alt="screenshot of comparison table" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="differential-analysis-of-exposures-between-groups">Differential analysis of exposures between groups<a class="anchor" aria-label="anchor" href="#differential-analysis-of-exposures-between-groups"></a>
</h3>
<p>The “Exposure Differential Analysis” tab is used to run differential
analysis on the signature exposures of annotated samples. There are 3
methods to perform the differential analysis: Wilcoxon Rank Sum Test,
Kruskal-Wallis Rank Sum Test, and a negative binomial regression
(glm).</p>
<p><img src="articles/ui_screenshots/DATab.jpg" alt="screenshot of differential analysis tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The Wilcoxon Rank Sum Test can be used to directly compare two groups
of samples. Below we display the Wilcoxon Rank Sum Test results between
LUAD and SKCM. Note that LUSC was ignored since it did not have a
specified pair.</p>
<p><img src="articles/ui_screenshots/DAOptions.jpg" alt="screenshot of differential analysis settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>From our table, we can see several of our signatures are
differentially exposed across tumor types. From our comparison analysis,
we can hypothesize that signature1 may be related to tobacco smoke, and
exposure to tobacco smoke occurs in lung cancers but not skin
cancers.</p>
<p><img src="articles/ui_screenshots/WilcoxonTable.jpg" alt="screenshot of differential analysis table" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="clustering-samples-based-on-exposures">Clustering samples based on exposures<a class="anchor" aria-label="anchor" href="#clustering-samples-based-on-exposures"></a>
</h3>
<p>The “Clustering” subtab provides several algorithms from the <a href="https://cran.r-project.org/web/packages/factoextra/index.html" class="external-link">factoextra</a>
and <a href="https://cran.r-project.org/web/packages/cluster/index.html" class="external-link">cluster</a>
packages to cluster samples based on exposure levels. After selecting
the musica result object, it is recommended to use the “Explore Number
of Clusters” box to help find the potential number of clusters in your
data. (Note: gap statistic may take a longer time to calculate.)</p>
<p><img src="articles/ui_screenshots/cluster1.jpg" alt="screenshot of clustering tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This line plot shows the within cluster sum of squares for each
number of clusters predicted using hierarchical clustering. The “elbow”
of the curve can be used as a starting place for the cluster number.</p>
<p><img src="articles/ui_screenshots/wss.jpg" alt="screenshot of within cluster sum of squares line plot" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The “Clustering” box is where you perform the clustering analysis. In
addition to clustering algorithm, several methods for calculating
dissimilarity matrix, imported from the <a href="https://cran.r-project.org/web/packages/philentropy/index.html" class="external-link">philentropy</a>
package, are also provided.</p>
<p><img src="articles/ui_screenshots/cluster2.jpg" alt="screenshot of clustering box settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This table is the output of clustering analysis, combined with
annotation.</p>
<p><img src="articles/ui_screenshots/cluster_tbl.jpg" alt="screenshot of clustering analysis output" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>In the “Visualization” box, users can make scatter plots to visualize
the clustering results on a UMAP panel. Three types of plots are
provided.</p>
<p><img src="articles/ui_screenshots/cluster_visual.jpg" alt="screenshot of visualization box settings" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “None” is selected, a single scatter plot will be made with points
colored by cluster label.</p>
<p><img src="articles/ui_screenshots/umap3.jpg" alt="screenshot of scatter plot" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “Signature” is selected, a subplot is made for each combination of
cluster label and signature. Points are colored by the level of the
specific signature.</p>
<p><img src="articles/ui_screenshots/umap1.jpg" alt="screenshot of subplots" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “Annotation” is selected, an additional select box will show up
and let you choose one type of user-supplied annotation of interest. A
subplot is generated for each combination of signature and category in
the annotation. Points are colored by cluster label.</p>
<p><img src="articles/ui_screenshots/umap2.jpg" alt="screenshot of subplots" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="heatmap-of-exposures">Heatmap of exposures<a class="anchor" aria-label="anchor" href="#heatmap-of-exposures"></a>
</h3>
<p>Heatmaps can be used to visualize the relative levels of all
signatures across tumors in a single plot. Select the result object you
want to use for heatmap visualization from the dropdown menu in the
start labeled ‘Select Result’. Choose from different settings and press
the ‘Plot’ button to visualize the heatmap. Check the ‘Proportional’
option to normalize the levels of signatures within a sample to sum to
one. Z-score normalization will scale each signature to have a mean of
zero and a standard deviation of one.. You can also show or hide the
tumor and signature names by checking or unchecking the ‘Show column
names’ and ‘Show row names’, respectively. If you want to display a
subset by signatures, click the ‘Selected Signatures’ option and select
the signatures to show. You can also select the ‘Samples’ and
‘Annotation’ options to subset by the available samples and
annotations.</p>
<p><img src="articles/ui_screenshots/heatmap.jpg" alt="screenshot of heatmap tab" style="border: 1px solid rgba(0,0,0,0.2)"><br><img src="articles/ui_screenshots/heatmap1.jpg" alt="screenshot of heatmap" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="download-results">Download results<a class="anchor" aria-label="anchor" href="#download-results"></a>
</h3>
<p>You can download your musica object by using the drop down menu. It
will be saved as .rds files which can be reuploaded back into the app at
a later time. Select the name of the musica object that click the
Download button to save the file to your computer.</p>
<p><img src="articles/ui_screenshots/download.jpg" alt="screenshot of download tab" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aaron Chevalier, Natasha Gurevich, Tao Guo, Joshua D. Campbell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
