<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of mutational signatures with the Shiny graphical user interface (GUI) • musicatk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of mutational signatures with the Shiny graphical user interface (GUI)">
<meta property="og:description" content="musicatk">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">musicatk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/musicatk" class="external-link">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/articles/tutorial_tcga_cnsl.html">Analysis in the R console</a>
    </li>
    <li>
      <a href="../articles/articles/tutorial_tcga_ui.html">Analysis with the user interface</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/campbio/musicatk" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of mutational signatures with the Shiny
graphical user interface (GUI)</h1>
                        <h4 data-toc-skip class="author">Nathan
Sahelijo, Zainab Khurshid, Tong Tong, Aaron Chevalier, Joshua
Campbell</h4>
            
            <h4 data-toc-skip class="date">Compiled May 03, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/campbio/musicatk/blob/HEAD/vignettes/articles/tutorial_tcga_ui.Rmd" class="external-link"><code>vignettes/articles/tutorial_tcga_ui.Rmd</code></a></small>
      <div class="hidden name"><code>tutorial_tcga_ui.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>A variety of exogenous exposures or endogenous biological processes
can contribute to the overall mutational load observed in human tumors.
Many different mutational patterns, or “mutational signatures”, have
been identified across different tumor types. These signatures can
provide a record of environmental exposure and can give clues about the
etiology of carcinogenesis. The Mutational Signature Comprehensive
Analysis Toolkit (musicatk) contains a complete end-to-end workflow for
characterization of mutational signatures in a cohort of samples.
musicatk has utilities for extracting variants from a variety of file
formats, multiple methods for discovery of novel signatures or
prediction of pre-existing signatures, and many types of downstream
visualizations for exploratory analysis. This package has the ability to
parse and combine multiple motif classes in the mutational signature
discovery or prediction processes. Mutation motifs include single base
substitutions (SBS), double base substitutions (DBS), insertions (INS)
and deletions (DEL). To start the app on your local machine, open
RStudio or R in the terminal and type the following in the console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">musicatk</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/musicatk.html">musicatk</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="importing-mutational-data">Importing mutational data<a class="anchor" aria-label="anchor" href="#importing-mutational-data"></a>
</h2>
<p>Users can upload their datasets through 1 of 3 different methods. -
“Import variants from files” - upload MAF and VCF files containing the
variant data. - “Import variants from TCGA” - easily import all open
access TCGA tumor datasets. - “Import previously generated musica
objects” - upload existing musica and result objects. <br></p>
<div class="section level3">
<h3 id="import-variants-from-files">Import variants from files<a class="anchor" aria-label="anchor" href="#import-variants-from-files"></a>
</h3>
<p>In order to discover or predict mutational signatures, we must first
set up our musica object by 1) extracting variants from files or objects
such as VCFs and MAFs, 2) selecting the appropriate reference genome 3)
creating a musica object, 4) adding sample-level annotations, and 5)
building a count tables for our variants of interest. Variants stored in
VCF or MAF files can be selected from your computation using the
‘browse’ button under the ‘Select Files’ label.</p>
<p><img src="ui_screenshots/import.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Clicking the ‘Add Samples’ button will add the files to the list of
samples to be imported. The delete and undo button next to the file can
be pressed to remove it or revert the change respectively. After adding
the files, the ‘Import’ button can be clicked to start the importing
process.</p>
<p><img src="ui_screenshots/import1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Once the process is completed, a notification in the bottom right
will appear. A variants table from the uploaded files will also appear
which can be downloaded by pressing the ‘Download Variants’ button.</p>
<p><img src="ui_screenshots/import2.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="import-tcga-datasets">Import TCGA Datasets<a class="anchor" aria-label="anchor" href="#import-tcga-datasets"></a>
</h3>
<p>This tab allows for selection of single or multiple datasets
generated in The Cancer Genome Atlas (TCGA). The original TCGA page for
the tumor name list can be referred to by selecting the ‘Full Tumor
List’ link on the top. Once the desired datasets have been selected, the
‘Import’ button at the bottom can be clicked to download the TCGA data.
It may take some time to download and process all of the selected
datasets. Once finished, a completion notification will appear at the
bottom right.</p>
<p><img src="ui_screenshots/TCGA.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="import-previously-generated-musica-objects">Import previously generated musica objects<a class="anchor" aria-label="anchor" href="#import-previously-generated-musica-objects"></a>
</h3>
<p>This tab can be used to import a user’s previously generated musica
variant object or a musica result object in .rda or .rds format. . The
browse button in the musica result or musica result object can be
selected for selecting files. Once selected, a bar saying file upload is
completed will appear. By default, the musica result objects will be
named after the file’s name but it can be changed in the text box
labeled “Name your musica result object”. Once finished, a notification
message in the bottom right will appear along with a table of variants.
If a musica variant object was uploaded then the users can skip the
Create Musica tab and move to the build tables tab directly. If a musica
result object was uploaded, then all the downstream tabs will be
accessible.</p>
<p><img src="ui_screenshots/upload_musica.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="creating-a-musica-object">Creating a musica object<a class="anchor" aria-label="anchor" href="#creating-a-musica-object"></a>
</h2>
<p>In this tab users can process variants and create a musica variant
object which can be used for discovery of signatures or prediction of
exposures. The reference genome can be selected from the drop down menu
labeled “Choose Genome” that matches the genome build of the variants
and the ‘Create Musica Object’ button can then be clicked. Once it’s
done, a completion notification message at the bottom right and a
variant summary table will appear. The musica object can be downloaded
by clicking the ‘Download Musica Object’ button and re-uploaded to save
time in subsequent sessions.</p>
<p><img src="ui_screenshots/create_musica.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br><img src="ui_screenshots/create_musica1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level2">
<h2 id="importing-sample-annotations">Importing sample annotations<a class="anchor" aria-label="anchor" href="#importing-sample-annotations"></a>
</h2>
<p>Sample annotations can be used to store information about each sample
such as tumor type or treatment status. They are optional and can be
used in downstream plotting functions such as <a href="../../docs/reference/plot_exposures.html"><em>plot_exposures</em></a>
or <a href="../../docs/reference/plot_umap.html"><em>plot_umap</em></a>
to group or color samples by a particular annotation.</p>
<p><img src="ui_screenshots/AnnotationsTab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Annotations are sample-level variables such as Tumor Type or Age.
These can be provided in a character delimited text file which must
include a column containing the sample IDs.</p>
<p><img src="ui_screenshots/annotation_file.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The sample IDs must match the sample names in the musica object. NA
values will be given to any samples not present in the annotation
file.</p>
<p><img src="ui_screenshots/AnnotationOptions.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br> After selecting the correct delimiter, a data table will appear
below to show the annotations that will be added to the musica object.
If your annotation file does not contain a header, unselect the “Header”
radio button. Choose the column that contains the sample names from the
“Sample Name Columns” dropdown and then click “Add Annotation”.</p>
<p><img src="ui_screenshots/AnnotationTable.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br> # Creating mutation count tables</p>
<div class="section level3">
<h3 id="create-standard-tables">Create standard tables<a class="anchor" aria-label="anchor" href="#create-standard-tables"></a>
</h3>
<p>The “Build Tables” tab generates count tables for different mutation
schemas. These schemas are the input to the mutational signature
discovery and prediction functions. To build the count tables, the user
must select 1 of the 5 standard motifs in the “Select Count Table”
dropdown.</p>
<p><br></p>
<p>SBS96 - Motifs are the six possible single base pair mutation types
times the four possibilities each for upstream and downstream context
bases (464 = 96 motifs) <br> SBS192_Trans - Motifs are an extension of
SBS96 multiplied by the transcriptional strand
(translated/untranslated), can be specified with “Transcript_Strand”.
<br> SBS192_Rep - Motifs are an extension of SBS96 multiplied by the
replication strand (leading/lagging), can be specified with
“Replication_Strand”. <br> DBS - Motifs are the 78 possible
double-base-pair substitutions. <br> INDEL - Motifs are 83 categories
intended to capture different categories of indels based on base-pair
change, repeats, or microhomology, insertion or deletion, and
length.</p>
<p><br></p>
<p>In addition to selecting a motif, the user must also provide the
reference genome.</p>
<p><img src="ui_screenshots/TableTab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="combine-tables">Combine tables<a class="anchor" aria-label="anchor" href="#combine-tables"></a>
</h3>
<p>Any two or more tables can be combined into a new table. The option
to build a combined table will only appear after the second table has
been created.</p>
<p><img src="ui_screenshots/combine-table.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-signatures-and-exposures">Estimating signatures and exposures<a class="anchor" aria-label="anchor" href="#estimating-signatures-and-exposures"></a>
</h2>
<p>The user has the option of discovering novel signatures and
exposures, or using previously discovered signatures to predict
exposures. To discover signatures use the “Discover” tab. To predict
signature exposures, use the “Predict” tab.</p>
<div class="section level3">
<h3 id="discovery-of-signatures-and-exposures">Discovery of signatures and exposures<a class="anchor" aria-label="anchor" href="#discovery-of-signatures-and-exposures"></a>
</h3>
<p>Mutational signatures and exposures are discovered using a Latent
Dirichlet Allocation (LDA) or a Non-Negative Matrix Factorization (NMF)
algorithms. These algorithms will deconvolute the mutation count matrix
into two matrices: 1) a “signature” matrix containing the probability of
each mutation type in each sample and 2) an “exposure” matrix containing
the estimated counts for each signature in each sample. Select a count
table, algorithm, number of signatures, and specify the name of the
result and then click “Discover signatures”.</p>
<p><img src="ui_screenshots/DiscoverTab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="predicting-exposures-from-existing-signatures">Predicting exposures from existing signatures<a class="anchor" aria-label="anchor" href="#predicting-exposures-from-existing-signatures"></a>
</h3>
<p>Exposures for samples can be predicted using an existing set of
signatures stored in a musica result object. Algorithms available for
prediction include “lda”, “decompTumor2Sig”, and “deconstructSigs”. Note
that “deconstructSigs” can only work with variants stored in hg19
format.</p>
<p><img src="ui_screenshots/PredictTab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The “Signatures to Predict” dropdown contains all the signatures in
the result object selected from the “Result to Predict” dropdown. You
can search this dropdown and select multiple signatures. In our example
we use the “lda” algorithm to predict exposures of 6 Cosmic
signatures.</p>
<p><img src="ui_screenshots/PredictDropdown.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="visualization-of-results">Visualization of results<a class="anchor" aria-label="anchor" href="#visualization-of-results"></a>
</h2>
<p>The data visualization tab can be used to generate customized plots
for signatures and exposures. Plots will be made interactive if the
“plotly” option is selected.</p>
<div class="section level3">
<h3 id="plot-signatures">Plot signatures<a class="anchor" aria-label="anchor" href="#plot-signatures"></a>
</h3>
<p>Bar plots can be used to display the probability of each type of
mutation within each signature.</p>
<p><img src="ui_screenshots/visual_sig.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Previously, 8 single-base signatures were identified from a cohort of
lung adenocarcinoma, lung squamous cell carcinoma, and skin cutaneous
melanoma samples from TCGA. By default, signatures are named by numbers,
but an option of renaming signatures is provided if you want to rename
them according to their possible etiology (e.g. Smoking) or closest
correlation to COSMIC (e.g. SBS4).</p>
<p><img src="ui_screenshots/sig_plot.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="plot-exposures">Plot exposures<a class="anchor" aria-label="anchor" href="#plot-exposures"></a>
</h3>
<p>Bar plots, box plots, violin plots and scatter plots can be used to
visualize the exposure levels of each signature in each sample.</p>
<p><img src="ui_screenshots/visual_exp.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>By default, a stacked bar plot sorted by the total number of
mutations is used. Each stacked bar shows the proportion of exposure of
each signature.</p>
<p><img src="ui_screenshots/exp_plot1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The stacked bar plot can be ordered by signatures. If Signatures is
selected in the Sort By option, a bucket list will show up to allow you
select the signatures you want to use by dragging them from the left box
to the right box. Users can also set a limit on the number of samples to
display.</p>
<p><img src="ui_screenshots/sorting.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This stacked bar plot is now ordered by the exposure of signature 1
and only the top 400 samples were included here.</p>
<p><img src="ui_screenshots/exp_plot2.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Box plot and violin plot can be used to visualize the distribution of
exposures or compare exposures between different groups of samples.</p>
<p><img src="ui_screenshots/exp_box_setting1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>By default, a box plot of exposure for each signature will be
shown.</p>
<p><img src="ui_screenshots/exp_plot3.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If an annotation file is provided, then you can visually compare the
exposure of signatures among different groups.</p>
<p><img src="ui_screenshots/exp_box_setting2.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>In this box plot, exposures of each signature were grouped by tumor
types. We can find that signatures 1 and 5 had high activities in lung
cancer samples, while signatures 3 and 8 had high activities in skin
cancer samples.</p>
<p><img src="ui_screenshots/exp_plot4.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>We can also group samples by signatures and then color by tumor
types.</p>
<p><img src="ui_screenshots/exp_box_setting3.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This plot can be used to visualize and compare the distribution of
exposure levels for signature across an annotation such as tumor
type.</p>
<p><img src="ui_screenshots/exp_plot5.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="downstream-analyses">Downstream analyses<a class="anchor" aria-label="anchor" href="#downstream-analyses"></a>
</h2>
<div class="section level3">
<h3 id="compare-signatures">Compare Signatures<a class="anchor" aria-label="anchor" href="#compare-signatures"></a>
</h3>
<p>A common analysis is to compare discovered signatures estimated in a
dataset to those generated in other datasets or to those in the <a href="https://cancer.sanger.ac.uk/cosmic/signatures" class="external-link">COSMIC
database</a>. The threshold acts as a cutoff similarity score and can be
any value between 0 and 1. Results will populate in a downloadable data
table.</p>
<p><img src="ui_screenshots/CompareTab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>Using the cosine metric with a threshold of 0.8,7 pairs of similar
signatures from COSMIC. Our Signature1 had a high cosine similarity to
Cosmic SBS4 indicating that some of our tumors have been exposed to
cigarette smoke.</p>
<p><img src="ui_screenshots/CompareTable.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="differential-analysis-of-exposures-between-groups">Differential analysis of exposures between groups<a class="anchor" aria-label="anchor" href="#differential-analysis-of-exposures-between-groups"></a>
</h3>
<p>The “Exposure Differential Analysis” tab is used to run differential
analysis on the signature exposures of annotated samples. There are 3
methods to perform the differential analysis: Wilcoxon Rank Sum Test,
Kruskal-Wallis Rank Sum Test, and a negative binomial regression
(glm).</p>
<p><img src="ui_screenshots/DATab.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The Wilcoxon Rank Sum Test can be used to directly compare two groups
of samples. Below we display the Wilcoxon Rank Sum Test results between
LUAD and SKCM. Note that LUSC was ignored since it did not have a
specified pair.</p>
<p><img src="ui_screenshots/DAOptions.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>From our table, we can see several of our signatures are
differentially exposed across tumor types. From our comparison analysis,
we can hypothesize that signature1 may be related to tobacco smoke, and
exposure to tobacco smoke occurs in lung cancers but not skin
cancers.</p>
<p><img src="ui_screenshots/WilcoxonTable.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="clustering-samples-based-on-exposures">Clustering samples based on exposures<a class="anchor" aria-label="anchor" href="#clustering-samples-based-on-exposures"></a>
</h3>
<p>The “Clustering” subtab provides several algorithms from the <a href="https://cran.r-project.org/web/packages/factoextra/index.html" class="external-link">factoextra</a>
and <a href="https://cran.r-project.org/web/packages/cluster/index.html" class="external-link">cluster</a>
packages to cluster samples based on exposure levels. After selecting
the musica result object, it is recommended to use the “Explore Number
of Clusters” box to help find the potential number of clusters in your
data. (Note: gap statistic may take a longer time to calculate.)</p>
<p><img src="ui_screenshots/cluster1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This line plot shows the within cluster sum of squares for each
number of clusters predicted using hierarchical clustering. The “elbow”
of the curve can be used as a starting place for the cluster number.</p>
<p><img src="ui_screenshots/wss.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>The “Clustering” box is where you perform the clustering analysis. In
addition to clustering algorithm, several methods for calculating
dissimilarity matrix, imported from the <a href="https://cran.r-project.org/web/packages/philentropy/index.html" class="external-link">philentropy</a>
package, are also provided.</p>
<p><img src="ui_screenshots/cluster2.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>This table is the output of clustering analysis, combined with
annotation.</p>
<p><img src="ui_screenshots/cluster_tbl.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>In the “Visualization” box, users can make scatter plots to visualize
the clustering results on a UMAP panel. Three types of plots are
provided.</p>
<p><img src="ui_screenshots/cluster_visual.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “None” is selected, a single scatter plot will be made with points
colored by cluster label.</p>
<p><img src="ui_screenshots/umap3.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “Signature” is selected, a subplot is made for each combination of
cluster label and signature. Points are colored by the level of the
specific signature.</p>
<p><img src="ui_screenshots/umap1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
<p>If “Annotation” is selected, an additional select box will show up
and let you choose one type of user-supplied annotation of interest. A
subplot is generated for each combination of signature and category in
the annotation. Points are colored by cluster label.</p>
<p><img src="ui_screenshots/umap2.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="heatmap-of-exposures">Heatmap of exposures<a class="anchor" aria-label="anchor" href="#heatmap-of-exposures"></a>
</h3>
<p>Heatmaps can be used to visualize the relative levels of all
signatures across tumors in a single plot. Select the result object you
want to use for heatmap visualization from the dropdown menu in the
start labeled ‘Select Result’. Choose from different settings and press
the ‘Plot’ button to visualize the heatmap. Check the ‘Proportional’
option to normalize the levels of signatures within a sample to sum to
one. Z-score normalization will scale each signature to have a mean of
zero and a standard deviation of one.. You can also show or hide the
tumor and signature names by checking or unchecking the ‘Show column
names’ and ‘Show row names’, respectively. If you want to display a
subset by signatures, click the ‘Selected Signatures’ option and select
the signatures to show. You can also select the ‘Samples’ and
‘Annotation’ options to subset by the available samples and
annotations.</p>
<p><img src="ui_screenshots/heatmap.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br><img src="ui_screenshots/heatmap1.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
<div class="section level3">
<h3 id="download-results">Download results<a class="anchor" aria-label="anchor" href="#download-results"></a>
</h3>
<p>You can download your musica object and musica result object from
using the respective drop down menus. They will be saved as .rds files
which can be reuploaded back into the app at a later time. Select the
name of the musica result or musica object that click the Download
button to save the file to your computer.</p>
<p><img src="ui_screenshots/download.jpg" style="border: 1px solid rgba(0,0,0,0.2)"><br></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Chevalier, Joshua D. Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
