<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Builds count tables using various mutation type schemas — build_standard_table • musicatk</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Builds count tables using various mutation type schemas — build_standard_table"><meta name="description" content='Generates count tables for different mutation type schemas which can be
used as input to the mutational signature discovery or prediction functions.
"SBS96" generates a table for single base substitutions following the
standard 96 mutation types derived from the trinucleotide context.
"SBS192" is the 96 mutation type schema with the addition of
transcriptional strand or replication strand information added to each base.
"DBS" generates a table for the double base substitution schema
used in COSMIC V3. "Indel" generates a table for insertions and
deletions following the schema used in COSMIC V3.'><meta property="og:description" content='Generates count tables for different mutation type schemas which can be
used as input to the mutational signature discovery or prediction functions.
"SBS96" generates a table for single base substitutions following the
standard 96 mutation types derived from the trinucleotide context.
"SBS192" is the 96 mutation type schema with the addition of
transcriptional strand or replication strand information added to each base.
"DBS" generates a table for the double base substitution schema
used in COSMIC V3. "Indel" generates a table for insertions and
deletions following the schema used in COSMIC V3.'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">musicatk</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="https://www.camplab.net/musicatk" aria-label="Home"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/tutorial_tcga_cnsl.html">Analysis in the R console</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial_tcga_ui.html">Analysis with the user interface</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/campbio/musicatk" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/camplab1" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Builds count tables using various mutation type schemas</h1>
      <small class="dont-index">Source: <a href="https://github.com/campbio/musicatk/blob/HEAD/R/standard_tables.R" class="external-link"><code>R/standard_tables.R</code></a></small>
      <div class="d-none name"><code>build_standard_table.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generates count tables for different mutation type schemas which can be
used as input to the mutational signature discovery or prediction functions.
<code>"SBS96"</code> generates a table for single base substitutions following the
standard 96 mutation types derived from the trinucleotide context.
<code>"SBS192"</code> is the 96 mutation type schema with the addition of
transcriptional strand or replication strand information added to each base.
<code>"DBS"</code> generates a table for the double base substitution schema
used in COSMIC V3. <code>"Indel"</code> generates a table for insertions and
deletions following the schema used in COSMIC V3.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">build_standard_table</span><span class="op">(</span></span>
<span>  <span class="va">musica</span>,</span>
<span>  <span class="va">g</span>,</span>
<span>  <span class="va">modality</span>,</span>
<span>  strand_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  table_name <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-musica">musica<a class="anchor" aria-label="anchor" href="#arg-musica"></a></dt>
<dd><p>A <code><a href="musica-class.html">musica</a></code> object.</p></dd>


<dt id="arg-g">g<a class="anchor" aria-label="anchor" href="#arg-g"></a></dt>
<dd><p>A BSgenome object indicating which genome
reference the variants and their coordinates were derived from.</p></dd>


<dt id="arg-modality">modality<a class="anchor" aria-label="anchor" href="#arg-modality"></a></dt>
<dd><p>Modality of table to build. One of <code>"SBS96"</code>,
<code>"SBS192"</code>, <code>"DBS"</code>, or <code>"Indel"</code>.</p></dd>


<dt id="arg-strand-type">strand_type<a class="anchor" aria-label="anchor" href="#arg-strand-type"></a></dt>
<dd><p>Strand type to use in SBS192 schema. One of
<code>"Transcript_Strand"</code> or <code>"Replication_Strand"</code>.
Only used if <code>modality = SBS192</code>.</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>If <code>TRUE</code>, any existing count table with the same
name will be overwritten. If <code>FALSE</code>, then an error will be thrown
if a table with the same name exists within the <code>musica</code> object.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Show progress bar for processed samples</p></dd>


<dt id="arg-table-name">table_name<a class="anchor" aria-label="anchor" href="#arg-table-name"></a></dt>
<dd><p>Use modality instead</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>No object will be returned. The count tables will be automatically
added to the <code>musica</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="select_genome.html">select_genome</a></span><span class="op">(</span><span class="st">"19"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">musica</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">build_standard_table</span><span class="op">(</span><span class="va">musica</span>, <span class="va">g</span>, <span class="st">"SBS96"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building count table from SBS with SBS96 schema</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Overwriting counts table: SBS96</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">musica</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="annotate_transcript_strand.html">annotate_transcript_strand</a></span><span class="op">(</span><span class="va">musica</span>, <span class="st">"19"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   403 genes were dropped because they have exons located on both strands</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   of the same reference sequence or on more than one reference sequence,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   so cannot be represented by a single genomic range.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Use 'single.strand.genes.only=FALSE' to get all the genes in a</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   GRangesList object, or use suppressMessages() to suppress this message.</span>
<span class="r-in"><span><span class="fu">build_standard_table</span><span class="op">(</span><span class="va">musica</span>, <span class="va">g</span>, <span class="st">"SBS192"</span>, <span class="st">"Transcript_Strand"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building count table from SBS and Transcript_Strand with SBS192 schema</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">musica</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">rep_range</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="annotate_replication_strand.html">annotate_replication_strand</a></span><span class="op">(</span><span class="va">musica</span>, <span class="va">rep_range</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">build_standard_table</span><span class="op">(</span><span class="va">musica</span>, <span class="va">g</span>, <span class="st">"SBS192"</span>, <span class="st">"Replication_Strand"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building count table from SBS and Replication_Strand with SBS192 schema</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dbs_musica</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">build_standard_table</span><span class="op">(</span><span class="va">dbs_musica</span>, <span class="va">g</span>, <span class="st">"DBS"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building count table from DBS with DBS78 schema</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Overwriting counts table: DBS78</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">indel_musica</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">build_standard_table</span><span class="op">(</span><span class="va">indel_musica</span>, <span class="va">g</span>, modality <span class="op">=</span> <span class="st">"INDEL"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building count table from INDELs with IND83 schema</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aaron Chevalier, Natasha Gurevich, Tao Guo, Joshua D. Campbell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

