<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Creates a musica object from a variant table — create_musica_from_variants • musicatk</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Creates a musica object from a variant table — create_musica_from_variants"><meta name="description" content="This function creates a musica object from a variant
table or matrix. The musica class stores variants information,
variant-level annotations, sample-level annotations, and count tables and
is used as input to the mutational signature discovery and prediction
algorithms. The input variant table or matrix must have columns for
chromosome, start position, end position, reference allele,
alternate allele, and sample names. The column names in the variant table
can be mapped using the chromosome_col, start_col,
end_col, ref_col, alt_col, and
sample_col parameters."><meta property="og:description" content="This function creates a musica object from a variant
table or matrix. The musica class stores variants information,
variant-level annotations, sample-level annotations, and count tables and
is used as input to the mutational signature discovery and prediction
algorithms. The input variant table or matrix must have columns for
chromosome, start position, end position, reference allele,
alternate allele, and sample names. The column names in the variant table
can be mapped using the chromosome_col, start_col,
end_col, ref_col, alt_col, and
sample_col parameters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">musicatk</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="https://www.camplab.net/musicatk" aria-label="Home"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/articles/installation.html">Installation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/articles/tutorial_tcga_cnsl.html">Analysis in the R console</a></li>
    <li><a class="dropdown-item" href="../articles/articles/tutorial_tcga_ui.html">Analysis with the user interface</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/campbio/musicatk" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/camplab1" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Creates a musica object from a variant table</h1>
      <small class="dont-index">Source: <a href="https://github.com/campbio/musicatk/blob/HEAD/R/load_data.R" class="external-link"><code>R/load_data.R</code></a></small>
      <div class="d-none name"><code>create_musica_from_variants.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function creates a <a href="musica-class.html">musica</a> object from a variant
table or matrix. The <a href="musica-class.html">musica</a> class stores variants information,
variant-level annotations, sample-level annotations, and count tables and
is used as input to the mutational signature discovery and prediction
algorithms. The input variant table or matrix must have columns for
chromosome, start position, end position, reference allele,
alternate allele, and sample names. The column names in the variant table
can be mapped using the <code>chromosome_col</code>, <code>start_col</code>,
<code>end_col</code>, <code>ref_col</code>, <code>alt_col</code>, and
<code>sample_col parameters</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_musica_from_variants</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">genome</span>,</span>
<span>  check_ref_chromosomes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check_ref_bases <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  chromosome_col <span class="op">=</span> <span class="st">"chr"</span>,</span>
<span>  start_col <span class="op">=</span> <span class="st">"start"</span>,</span>
<span>  end_col <span class="op">=</span> <span class="st">"end"</span>,</span>
<span>  ref_col <span class="op">=</span> <span class="st">"ref"</span>,</span>
<span>  alt_col <span class="op">=</span> <span class="st">"alt"</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  extra_fields <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  standardize_indels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  convert_dbs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A data.table, matrix, or data.frame that contains columns with
the variant information.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>A BSgenome object indicating which genome
reference the variants and their coordinates were derived from.</p></dd>


<dt id="arg-check-ref-chromosomes">check_ref_chromosomes<a class="anchor" aria-label="anchor" href="#arg-check-ref-chromosomes"></a></dt>
<dd><p>Whether to peform a check to ensure that
the chromosomes in the <code>variant</code> object match the reference
chromosomes in the <code>genome</code> object. If there are mismatches, this
may cause errors in downstream generation of count tables. If mismatches
occur, an attept to be automatically fix these with the
<code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></code> function will be made.
Default <code>TRUE</code>.</p></dd>


<dt id="arg-check-ref-bases">check_ref_bases<a class="anchor" aria-label="anchor" href="#arg-check-ref-bases"></a></dt>
<dd><p>Whether to check if the reference bases in the
<code>variant</code> object match the reference bases in the <code>genome</code>
object. Default <code>TRUE</code>.</p></dd>


<dt id="arg-chromosome-col">chromosome_col<a class="anchor" aria-label="anchor" href="#arg-chromosome-col"></a></dt>
<dd><p>The name of the column that contains the chromosome
reference for each variant. Default <code>"chr"</code>.</p></dd>


<dt id="arg-start-col">start_col<a class="anchor" aria-label="anchor" href="#arg-start-col"></a></dt>
<dd><p>The name of the column that contains the start
position for each variant. Default <code>"start"</code>.</p></dd>


<dt id="arg-end-col">end_col<a class="anchor" aria-label="anchor" href="#arg-end-col"></a></dt>
<dd><p>The name of the column that contains the end
position for each variant. Default <code>"end"</code>.</p></dd>


<dt id="arg-ref-col">ref_col<a class="anchor" aria-label="anchor" href="#arg-ref-col"></a></dt>
<dd><p>The name of the column that contains the reference
base(s) for each variant. Default <code>"ref"</code>.</p></dd>


<dt id="arg-alt-col">alt_col<a class="anchor" aria-label="anchor" href="#arg-alt-col"></a></dt>
<dd><p>The name of the column that contains the alternative
base(s) for each variant. Default <code>"alt"</code>.</p></dd>


<dt id="arg-sample-col">sample_col<a class="anchor" aria-label="anchor" href="#arg-sample-col"></a></dt>
<dd><p>The name of the column that contains the sample
id for each variant. Default <code>"sample"</code>.</p></dd>


<dt id="arg-extra-fields">extra_fields<a class="anchor" aria-label="anchor" href="#arg-extra-fields"></a></dt>
<dd><p>Which additional fields to extract and include in
the musica object. Default <code>NULL</code>.</p></dd>


<dt id="arg-standardize-indels">standardize_indels<a class="anchor" aria-label="anchor" href="#arg-standardize-indels"></a></dt>
<dd><p>Flag to convert indel style (e.g. `C &gt; CAT`
becomes `- &gt; AT` and `GCACA &gt; G` becomes `CACA &gt; -`)</p></dd>


<dt id="arg-convert-dbs">convert_dbs<a class="anchor" aria-label="anchor" href="#arg-convert-dbs"></a></dt>
<dd><p>Flag to convert adjacent SBS into DBS (original SBS are
removed)</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print status messages during error checking.
Default <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a musica object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">maf_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"public_TCGA.LUSC.maf"</span>,</span></span>
<span class="r-in"><span>  package <span class="op">=</span> <span class="st">"musicatk"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="extract_variants_from_maf_file.html">extract_variants_from_maf_file</a></span><span class="op">(</span><span class="va">maf_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="select_genome.html">select_genome</a></span><span class="op">(</span><span class="st">"38"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">musica</span> <span class="op">&lt;-</span> <span class="fu">create_musica_from_variants</span><span class="op">(</span>x <span class="op">=</span> <span class="va">variants</span>, genome <span class="op">=</span> <span class="va">g</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking that chromosomes in the 'variant' object match chromosomes in the 'genome' object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Checking that the reference bases in the 'variant' object match the reference bases in the 'genome' object.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardizing INS/DEL style</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting adjacent SBS into DBS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 4 SBS converted to DBS</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aaron Chevalier, Natasha Gurevich, Tao Guo, Joshua D. Campbell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

