<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ musicatk</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">musicatk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/musicatk" class="external-link">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/articles/tutorial_tcga_cnsl.html">Analysis in the R console</a>
    </li>
    <li>
      <a href="../articles/articles/tutorial_tcga_ui.html">Analysis with the user interface</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/campbio/musicatk" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="importing-and-processing">Importing and processing <a href="#importing-and-processing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to import, extract, and process variants</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="extract_variants.html">extract_variants()</a></code> </p>
        </td>
        <td><p>Extract variants from mutliple objects</p></td>
      </tr><tr><td>
          <p><code><a href="extract_variants_from_maf.html">extract_variants_from_maf()</a></code> </p>
        </td>
        <td><p>Extract variants from a maf object</p></td>
      </tr><tr><td>
          <p><code><a href="extract_variants_from_maf_file.html">extract_variants_from_maf_file()</a></code> </p>
        </td>
        <td><p>Extracts variants from a maf file</p></td>
      </tr><tr><td>
          <p><code><a href="extract_variants_from_matrix.html">extract_variants_from_matrix()</a></code> </p>
        </td>
        <td><p>Extract variants from matrix or data.frame like objects</p></td>
      </tr><tr><td>
          <p><code><a href="extract_variants_from_vcf.html">extract_variants_from_vcf()</a></code> </p>
        </td>
        <td><p>Extracts variants from a VariantAnnotation VCF object</p></td>
      </tr><tr><td>
          <p><code><a href="extract_variants_from_vcf_file.html">extract_variants_from_vcf_file()</a></code> </p>
        </td>
        <td><p>Extracts variants from a vcf file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="musica-objects">musica objects <a href="#musica-objects" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to create and subset musica objects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="select_genome.html">select_genome()</a></code> </p>
        </td>
        <td><p>Helper function to load common human or mouse genomes</p></td>
      </tr><tr><td>
          <p><code><a href="create_musica.html">create_musica()</a></code> </p>
        </td>
        <td><p>Creates a musica object from a variant table</p></td>
      </tr><tr><td>
          <p><code><a href="subset_musica_by_annotation.html">subset_musica_by_annotation()</a></code> </p>
        </td>
        <td><p>Creates a new musica object subsetted to only one value of a sample annotation</p></td>
      </tr><tr><td>
          <p><code><a href="subset_musica_by_counts.html">subset_musica_by_counts()</a></code> </p>
        </td>
        <td><p>Creates a new musica subsetted to only samples with enough variants</p></td>
      </tr><tr><td>
          <p><code><a href="subset_variant_by_type.html">subset_variant_by_type()</a></code> </p>
        </td>
        <td><p>Subsets a variant table based on Variant Type</p></td>
      </tr><tr><td>
          <p><code><a href="subset_variants_by_samples.html">subset_variants_by_samples()</a></code> </p>
        </td>
        <td><p>Return sample from musica_variant object</p></td>
      </tr><tr><td>
          <p><code><a href="variants.html">variants()</a></code> <code><a href="variants.html">`variants&lt;-`()</a></code> </p>
        </td>
        <td><p>Retrieve variants from a musica or musica_result object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="creating-mutation-count-tables">Creating mutation count tables <a href="#creating-mutation-count-tables" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to build and combine mutation count tables</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="build_standard_table.html">build_standard_table()</a></code> </p>
        </td>
        <td><p>Builds count tables using various mutation type schemas</p></td>
      </tr><tr><td>
          <p><code><a href="build_custom_table.html">build_custom_table()</a></code> </p>
        </td>
        <td><p>Builds a custom table from specified user variants</p></td>
      </tr><tr><td>
          <p><code><a href="add_flank_to_variants.html">add_flank_to_variants()</a></code> </p>
        </td>
        <td><p>Uses a genome object to find context and add it to the variant table</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_replication_strand.html">annotate_replication_strand()</a></code> </p>
        </td>
        <td><p>Add replication strand annotation to SBS variants based on bedgraph file</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_transcript_strand.html">annotate_transcript_strand()</a></code> </p>
        </td>
        <td><p>Add transcript strand annotation to SBS variants (defined in genes only)</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_variant_length.html">annotate_variant_length()</a></code> </p>
        </td>
        <td><p>Adds an annotation to the input musica's variant table with length of each
variant</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_variant_type.html">annotate_variant_type()</a></code> </p>
        </td>
        <td><p>Annotate variants with variant type ("SBS", "INS", "DEl", "DBS")</p></td>
      </tr><tr><td>
          <p><code><a href="combine_count_tables.html">combine_count_tables()</a></code> </p>
        </td>
        <td><p>Combines tables into a single table that can be used for discovery/prediction</p></td>
      </tr><tr><td>
          <p><code><a href="extract_count_tables.html">extract_count_tables()</a></code> </p>
        </td>
        <td><p>Extract count tables list from a musica object</p></td>
      </tr><tr><td>
          <p><code><a href="tables.html">tables()</a></code> <code><a href="tables.html">`tables&lt;-`()</a></code> </p>
        </td>
        <td><p>Retrieve the list of count_tables from a musica or musica_result 
object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="sample-annotations">Sample annotations <a href="#sample-annotations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to add and manipulate sample annotations</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="samp_annot.html">samp_annot()</a></code> <code><a href="samp_annot.html">`samp_annot&lt;-`()</a></code> </p>
        </td>
        <td><p>Get or set sample annotations from a musica or musica_result object</p></td>
      </tr><tr><td>
          <p><code><a href="sample_names.html">sample_names()</a></code> </p>
        </td>
        <td><p>Retrieve sample names from a musica or musica_result object</p></td>
      </tr><tr><td>
          <p><code><a href="drop_annotation.html">drop_annotation()</a></code> </p>
        </td>
        <td><p>Drops a column from the variant table that the user no longer needs</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="discovery-and-prediction">Discovery and prediction <a href="#discovery-and-prediction" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to identify signatures and exposures from mutation count tables</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="discover_signatures.html">discover_signatures()</a></code> </p>
        </td>
        <td><p>Discover mutational signatures</p></td>
      </tr><tr><td>
          <p><code><a href="exposures.html">exposures()</a></code> <code><a href="exposures.html">`exposures&lt;-`()</a></code> </p>
        </td>
        <td><p>Retrieve exposures from a musica_result object</p></td>
      </tr><tr><td>
          <p><code><a href="signatures.html">signatures()</a></code> <code><a href="signatures.html">`signatures&lt;-`()</a></code> </p>
        </td>
        <td><p>Retrieve signatures from a musica_result object</p></td>
      </tr><tr><td>
          <p><code><a href="name_signatures.html">name_signatures()</a></code> </p>
        </td>
        <td><p>Return sample from musica object</p></td>
      </tr><tr><td>
          <p><code><a href="auto_predict_grid.html">auto_predict_grid()</a></code> </p>
        </td>
        <td><p>Automatic filtering of signatures for exposure prediction gridded across
specific annotation</p></td>
      </tr><tr><td>
          <p><code><a href="combine_predict_grid.html">combine_predict_grid()</a></code> </p>
        </td>
        <td><p>Combine prediction grid list into a result object. Exposure values are zero
for samples in an annotation where that signature was not predicted</p></td>
      </tr><tr><td>
          <p><code><a href="generate_result_grid.html">generate_result_grid()</a></code> </p>
        </td>
        <td><p>Generate result_grid from musica based on annotation and range of k</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to identify signatures and exposures from mutation count tables</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plot_exposures.html">plot_exposures()</a></code> </p>
        </td>
        <td><p>Display sample exposures with bar, box, or violin plots</p></td>
      </tr><tr><td>
          <p><code><a href="plot_signatures.html">plot_signatures()</a></code> </p>
        </td>
        <td><p>Plots the mutational signatures</p></td>
      </tr><tr><td>
          <p><code><a href="plot_umap.html">plot_umap()</a></code> </p>
        </td>
        <td><p>Plot a UMAP from a musica result</p></td>
      </tr><tr><td>
          <p><code><a href="plot_cluster.html">plot_cluster()</a></code> </p>
        </td>
        <td><p>Visualize clustering results</p></td>
      </tr><tr><td>
          <p><code><a href="plot_differential_analysis.html">plot_differential_analysis()</a></code> </p>
        </td>
        <td><p>Compare exposures of annotated samples</p></td>
      </tr><tr><td>
          <p><code><a href="plot_heatmap.html">plot_heatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmaps using the exposures matrix</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sample_counts.html">plot_sample_counts()</a></code> </p>
        </td>
        <td><p>Plot distribution of sample counts</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sample_reconstruction_error.html">plot_sample_reconstruction_error()</a></code> </p>
        </td>
        <td><p>Plot reconstruction error for a sample</p></td>
      </tr><tr><td>
          <p><code><a href="predict_exposure.html">predict_exposure()</a></code> </p>
        </td>
        <td><p>Prediction of exposures in new samples using pre-existing signatures</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="downstream-analysis">Downstream analysis <a href="#downstream-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to compare the signatures between two results objects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="create_umap.html">create_umap()</a></code> </p>
        </td>
        <td><p>Create a UMAP from a musica result</p></td>
      </tr><tr><td>
          <p><code><a href="umap.html">umap()</a></code> <code><a href="umap.html">`umap&lt;-`()</a></code> </p>
        </td>
        <td><p>Retrieve umap list from a musica_result object</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_exposure.html">cluster_exposure()</a></code> </p>
        </td>
        <td><p>Perform clustering analysis from a musica result object</p></td>
      </tr><tr><td>
          <p><code><a href="k_select.html">k_select()</a></code> </p>
        </td>
        <td><p>Plots for helping decide number of clusters</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="signature-comparisons">Signature comparisons <a href="#signature-comparisons" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to compare the signatures between two results objects</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="compare_cosmic_v2.html">compare_cosmic_v2()</a></code> </p>
        </td>
        <td><p>Compare a result object to COSMIC V2 SBS Signatures (combination whole-exome
and whole-genome)</p></td>
      </tr><tr><td>
          <p><code><a href="compare_cosmic_v3.html">compare_cosmic_v3()</a></code> </p>
        </td>
        <td><p>Compare a result object to COSMIC V3 Signatures; Select exome or genome for
SBS and only genome for DBS or Indel classes</p></td>
      </tr><tr><td>
          <p><code><a href="compare_results.html">compare_results()</a></code> </p>
        </td>
        <td><p>Compare two result files to find similar signatures</p></td>
      </tr><tr><td>
          <p><code><a href="exposure_differential_analysis.html">exposure_differential_analysis()</a></code> </p>
        </td>
        <td><p>Compare exposures of annotated samples</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="cosmic-signatures">COSMIC signatures <a href="#cosmic-signatures" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Result objects that contain COSMIC signatures</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cosmic_v2_sigs.html">cosmic_v2_sigs</a></code> </p>
        </td>
        <td><p>COSMIC v2 SBS96 Signatures Result Object</p></td>
      </tr><tr><td>
          <p><code><a href="cosmic_v3_dbs_sigs.html">cosmic_v3_dbs_sigs</a></code> </p>
        </td>
        <td><p>COSMIC v3 DBS Genome Signatures Result Object</p></td>
      </tr><tr><td>
          <p><code><a href="cosmic_v3_indel_sigs.html">cosmic_v3_indel_sigs</a></code> </p>
        </td>
        <td><p>COSMIC v3 Indel Genome Signatures Result Object</p></td>
      </tr><tr><td>
          <p><code><a href="cosmic_v3_sbs_sigs.html">cosmic_v3_sbs_sigs</a></code> </p>
        </td>
        <td><p>COSMIC v3 SBS96 Genome Signatures Result Object</p></td>
      </tr><tr><td>
          <p><code><a href="cosmic_v3_sbs_sigs_exome.html">cosmic_v3_sbs_sigs_exome</a></code> </p>
        </td>
        <td><p>COSMIC v3 SBS96 Exome Signatures Result Object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="miscellaneous-functions">Miscellaneous functions <a href="#miscellaneous-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cosmic_v2_subtype_map.html">cosmic_v2_subtype_map()</a></code> </p>
        </td>
        <td><p>Input a cancer subtype to return a list of related COSMIC signatures</p></td>
      </tr><tr><td>
          <p><code><a href="get_musica.html">get_musica()</a></code> </p>
        </td>
        <td><p>Retrieve musica from a musica_result object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-objects-and-classes">Data objects and classes <a href="#data-objects-and-classes" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="musica_sbs96.html">musica_sbs96</a></code> </p>
        </td>
        <td><p>musica_sbs96</p></td>
      </tr><tr><td>
          <p><code><a href="musica_sbs96_tiny.html">musica_sbs96_tiny</a></code> </p>
        </td>
        <td><p>musica_sbs96_tiny</p></td>
      </tr><tr><td>
          <p><code><a href="dbs_musica.html">dbs_musica</a></code> </p>
        </td>
        <td><p>dbs_musica</p></td>
      </tr><tr><td>
          <p><code><a href="indel_musica.html">indel_musica</a></code> </p>
        </td>
        <td><p>indel_musica</p></td>
      </tr><tr><td>
          <p><code><a href="musica.html">musica</a></code> </p>
        </td>
        <td><p>musica</p></td>
      </tr><tr><td>
          <p><code><a href="musica_annot.html">musica_annot</a></code> </p>
        </td>
        <td><p>musica_annot</p></td>
      </tr><tr><td>
          <p><code><a href="res.html">res</a></code> </p>
        </td>
        <td><p>res</p></td>
      </tr><tr><td>
          <p><code><a href="res_annot.html">res_annot</a></code> </p>
        </td>
        <td><p>res_annot</p></td>
      </tr><tr><td>
          <p><code><a href="rep_range.html">rep_range</a></code> </p>
        </td>
        <td><p>Replication Timing Data as GRanges Object</p></td>
      </tr><tr><td>
          <p><code><a href="musica-class.html">musica-class</a></code> </p>
        </td>
        <td><p>The primary object that contains variants, count_tables,
and samples annotations</p></td>
      </tr><tr><td>
          <p><code><a href="count_table-class.html">count_table-class</a></code> </p>
        </td>
        <td><p>Object containing the count table matrices, their names and descriptions
that we generated by provided and by user functions. These are used to
discover and infer signatures and exposures.</p></td>
      </tr><tr><td>
          <p><code><a href="musica_result-class.html">musica_result-class</a></code> </p>
        </td>
        <td><p>Object containing deconvolved/predicted signatures, sample weights, and
the musica object the result was generated from</p></td>
      </tr><tr><td>
          <p><code><a href="musica_result_grid-class.html">musica_result_grid-class</a></code> </p>
        </td>
        <td><p>Object containing the result objects generated from the combination of
annotations and a range of k values</p></td>
      </tr><tr><td>
          <p><code><a href="built_tables.html">built_tables()</a></code> </p>
        </td>
        <td><p>Retrieve the names of count_tables from a musica or musica_result 
object</p></td>
      </tr><tr><td>
          <p><code><a href="musicatk.html">musicatk()</a></code> </p>
        </td>
        <td><p>Starts the musicatk interactive Shiny app</p></td>
      </tr><tr><td>
          <p><code><a href="rc.html">rc()</a></code> </p>
        </td>
        <td><p>Reverse complement of a string using biostrings</p></td>
      </tr><tr><td>
          <p><code><a href="table_selected.html">table_selected()</a></code> </p>
        </td>
        <td><p>Retrieve table name used for plotting from a musica_result object</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Aaron Chevalier, Joshua D. Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

