<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ musicatk</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">musicatk</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="https://www.camplab.net/musicatk" aria-label="Home"><span class="fa fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/installation.html">Installation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/tutorial_tcga_cnsl.html">Analysis in the R console</a></li>
    <li><a class="dropdown-item" href="../articles/tutorial_tcga_ui.html">Analysis with the user interface</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/campbio/musicatk" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/camplab1" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="importing-and-processing">Importing and processing<a class="anchor" aria-label="anchor" href="#importing-and-processing"></a></h2>

      <div class="section-desc"><p>Functions to import, extract, and process variants</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="extract_variants.html">extract_variants()</a></code>

        </dt>
        <dd>Extract variants from mutliple objects</dd>
      </dl><dl><dt>

          <code><a href="extract_variants_from_maf.html">extract_variants_from_maf()</a></code>

        </dt>
        <dd>Extract variants from a maf object</dd>
      </dl><dl><dt>

          <code><a href="extract_variants_from_maf_file.html">extract_variants_from_maf_file()</a></code>

        </dt>
        <dd>Extracts variants from a maf file</dd>
      </dl><dl><dt>

          <code><a href="extract_variants_from_matrix.html">extract_variants_from_matrix()</a></code>

        </dt>
        <dd>Extract variants from matrix or data.frame like objects</dd>
      </dl><dl><dt>

          <code><a href="extract_variants_from_vcf.html">extract_variants_from_vcf()</a></code>

        </dt>
        <dd>Extracts variants from a VariantAnnotation VCF object</dd>
      </dl><dl><dt>

          <code><a href="extract_variants_from_vcf_file.html">extract_variants_from_vcf_file()</a></code>

        </dt>
        <dd>Extracts variants from a vcf file</dd>
      </dl></div><div class="section level2">
      <h2 id="musica-objects">musica objects<a class="anchor" aria-label="anchor" href="#musica-objects"></a></h2>

      <div class="section-desc"><p>Functions to create and subset musica objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="select_genome.html">select_genome()</a></code>

        </dt>
        <dd>Helper function to load common human or mouse genomes</dd>
      </dl><dl><dt>

          <code><a href="create_musica_from_variants.html">create_musica_from_variants()</a></code>

        </dt>
        <dd>Creates a musica object from a variant table</dd>
      </dl><dl><dt>

          <code><a href="create_musica_from_counts.html">create_musica_from_counts()</a></code>

        </dt>
        <dd>Creates a musica object from a mutation count table</dd>
      </dl><dl><dt>

          <code><a href="subset_musica_by_annotation.html">subset_musica_by_annotation()</a></code>

        </dt>
        <dd>Creates a new musica object subsetted to only one value of a sample annotation</dd>
      </dl><dl><dt>

          <code><a href="subset_musica_by_counts.html">subset_musica_by_counts()</a></code>

        </dt>
        <dd>Creates a new musica subsetted to only samples with enough variants</dd>
      </dl><dl><dt>

          <code><a href="subset_variant_by_type.html">subset_variant_by_type()</a></code>

        </dt>
        <dd>Subsets a variant table based on Variant Type</dd>
      </dl><dl><dt>

          <code><a href="subset_variants_by_samples.html">subset_variants_by_samples()</a></code>

        </dt>
        <dd>Return sample from musica_variant object</dd>
      </dl><dl><dt>

          <code><a href="variants.html">variants()</a></code> <code><a href="variants.html">`variants&lt;-`()</a></code>

        </dt>
        <dd>Retrieve variants from a musica object</dd>
      </dl></div><div class="section level2">
      <h2 id="creating-mutation-count-tables">Creating mutation count tables<a class="anchor" aria-label="anchor" href="#creating-mutation-count-tables"></a></h2>

      <div class="section-desc"><p>Functions to build and combine mutation count tables</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="build_standard_table.html">build_standard_table()</a></code>

        </dt>
        <dd>Builds count tables using various mutation type schemas</dd>
      </dl><dl><dt>

          <code><a href="build_custom_table.html">build_custom_table()</a></code>

        </dt>
        <dd>Builds a custom table from specified user variants</dd>
      </dl><dl><dt>

          <code><a href="add_flank_to_variants.html">add_flank_to_variants()</a></code>

        </dt>
        <dd>Uses a genome object to find context and add it to the variant table</dd>
      </dl><dl><dt>

          <code><a href="annotate_replication_strand.html">annotate_replication_strand()</a></code>

        </dt>
        <dd>Add replication strand annotation to SBS variants based on bedgraph file</dd>
      </dl><dl><dt>

          <code><a href="annotate_transcript_strand.html">annotate_transcript_strand()</a></code>

        </dt>
        <dd>Add transcript strand annotation to SBS variants (defined in genes only)</dd>
      </dl><dl><dt>

          <code><a href="annotate_variant_length.html">annotate_variant_length()</a></code>

        </dt>
        <dd>Adds an annotation to the input musica's variant table with length of each variant</dd>
      </dl><dl><dt>

          <code><a href="annotate_variant_type.html">annotate_variant_type()</a></code>

        </dt>
        <dd>Annotate variants with variant type ("SBS", "INS", "DEl", "DBS")</dd>
      </dl><dl><dt>

          <code><a href="combine_count_tables.html">combine_count_tables()</a></code>

        </dt>
        <dd>Combines tables into a single table that can be used for discovery/prediction</dd>
      </dl><dl><dt>

          <code><a href="extract_count_tables.html">extract_count_tables()</a></code>

        </dt>
        <dd>Extract count tables list from a musica object</dd>
      </dl><dl><dt>

          <code><a href="tables.html">tables()</a></code> <code><a href="tables.html">`tables&lt;-`()</a></code>

        </dt>
        <dd>Retrieve the list of count_tables from a musica object</dd>
      </dl><dl><dt>

          <code><a href="get_count_table.html">get_count_table()</a></code>

        </dt>
        <dd>Retrieve count_table matrix from count_table object</dd>
      </dl></div><div class="section level2">
      <h2 id="sample-annotations">Sample annotations<a class="anchor" aria-label="anchor" href="#sample-annotations"></a></h2>

      <div class="section-desc"><p>Functions to add and manipulate sample annotations</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="samp_annot.html">samp_annot()</a></code> <code><a href="samp_annot.html">`samp_annot&lt;-`()</a></code>

        </dt>
        <dd>Get or set sample annotations from a musica object</dd>
      </dl><dl><dt>

          <code><a href="sample_names.html">sample_names()</a></code>

        </dt>
        <dd>Retrieve sample names from a musica object</dd>
      </dl><dl><dt>

          <code><a href="drop_annotation.html">drop_annotation()</a></code>

        </dt>
        <dd>Drops a column from the variant table that the user no longer needs</dd>
      </dl></div><div class="section level2">
      <h2 id="k-value-help">K value help<a class="anchor" aria-label="anchor" href="#k-value-help"></a></h2>

      <div class="section-desc"><p>Functions to help determine k value</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="compare_k_vals.html">compare_k_vals()</a></code>

        </dt>
        <dd>Compare k values</dd>
      </dl><dl><dt>

          <code><a href="plot_k_comparison.html">plot_k_comparison()</a></code>

        </dt>
        <dd>Plot k comparison</dd>
      </dl></div><div class="section level2">
      <h2 id="discovery-and-prediction">Discovery and prediction<a class="anchor" aria-label="anchor" href="#discovery-and-prediction"></a></h2>

      <div class="section-desc"><p>Functions to identify signatures and exposures from mutation count tables</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="discover_signatures.html">discover_signatures()</a></code>

        </dt>
        <dd>Discover mutational signatures</dd>
      </dl><dl><dt>

          <code><a href="exposures.html">exposures()</a></code> <code><a href="exposures.html">`exposures&lt;-`()</a></code>

        </dt>
        <dd>Retrieve exposures from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="signatures.html">signatures()</a></code> <code><a href="signatures.html">`signatures&lt;-`()</a></code>

        </dt>
        <dd>Retrieve signatures from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="name_signatures.html">name_signatures()</a></code>

        </dt>
        <dd>Rename signatures for a model</dd>
      </dl><dl><dt>

          <code><a href="auto_predict_grid.html">auto_predict_grid()</a></code>

        </dt>
        <dd>Automatic filtering of signatures for exposure prediction gridded across specific annotation</dd>
      </dl><dl><dt>

          <code><a href="combine_predict_grid.html">combine_predict_grid()</a></code>

        </dt>
        <dd>Combine signatures and exposures of different models. Exposure values are zero for samples in an annotation where that signature was not predicted</dd>
      </dl><dl><dt>

          <code><a href="generate_result_grid.html">generate_result_grid()</a></code>

        </dt>
        <dd>Generate result_grid from musica based on annotation and range of k</dd>
      </dl></div><div class="section level2">
      <h2 id="accessing-result-data-in-musica-objects">Accessing result data in musica objects<a class="anchor" aria-label="anchor" href="#accessing-result-data-in-musica-objects"></a></h2>

      <div class="section-desc"><p>Functions to add, extract, and access data within the result_list of a musica object</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_result.html">add_result()</a></code>

        </dt>
        <dd>Load an external model into a musica object</dd>
      </dl><dl><dt>

          <code><a href="get_model.html">get_model()</a></code>

        </dt>
        <dd>Retrieve model from a musica or result collection object</dd>
      </dl><dl><dt>

          <code><a href="get_result_list_entry.html">get_result_list_entry()</a></code>

        </dt>
        <dd>Retrieve result_list entry from a musica object</dd>
      </dl><dl><dt>

          <code><a href="get_modality.html">get_modality()</a></code>

        </dt>
        <dd>Retrieve a specific modality entry from a musica or result_collection object</dd>
      </dl><dl><dt>

          <code><a href="result_list.html">result_list()</a></code> <code><a href="result_list.html">`result_list&lt;-`()</a></code>

        </dt>
        <dd>Retrieve result_list from a musica object</dd>
      </dl><dl><dt>

          <code><a href="hyperparameter.html">hyperparameter()</a></code> <code><a href="hyperparameter.html">`hyperparameter&lt;-`()</a></code>

        </dt>
        <dd>Retrieve hyperparameter from a musica or result_collection object</dd>
      </dl><dl><dt>

          <code><a href="parameter.html">parameter()</a></code> <code><a href="parameter.html">`parameter&lt;-`()</a></code>

        </dt>
        <dd>Retrieve parameter from a musica or result_collection object</dd>
      </dl><dl><dt>

          <code><a href="signatures.html">signatures()</a></code> <code><a href="signatures.html">`signatures&lt;-`()</a></code>

        </dt>
        <dd>Retrieve signatures from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="exposures.html">exposures()</a></code> <code><a href="exposures.html">`exposures&lt;-`()</a></code>

        </dt>
        <dd>Retrieve exposures from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="num_signatures.html">num_signatures()</a></code> <code><a href="num_signatures.html">`num_signatures&lt;-`()</a></code>

        </dt>
        <dd>Retrieve num_signatures from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="other_parameters.html">other_parameters()</a></code> <code><a href="other_parameters.html">`other_parameters&lt;-`()</a></code>

        </dt>
        <dd>Retrieve other_parameters from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="credible_intervals.html">credible_intervals()</a></code> <code><a href="credible_intervals.html">`credible_intervals&lt;-`()</a></code>

        </dt>
        <dd>Retrieve credible_intervals from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="metrics.html">metrics()</a></code> <code><a href="metrics.html">`metrics&lt;-`()</a></code>

        </dt>
        <dd>Retrieve metrics from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="model_id.html">model_id()</a></code> <code><a href="model_id.html">`model_id&lt;-`()</a></code>

        </dt>
        <dd>Retrieve model_id from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="modality.html">modality()</a></code> <code><a href="modality.html">`modality&lt;-`()</a></code>

        </dt>
        <dd>Retrieve modality from a result_model, result_collection, or musica object</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>

      <div class="section-desc"><p>Functions to identify signatures and exposures from mutation count tables</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plot_exposures.html">plot_exposures()</a></code>

        </dt>
        <dd>Display sample exposures with bar, box, or violin plots</dd>
      </dl><dl><dt>

          <code><a href="plot_signatures.html">plot_signatures()</a></code>

        </dt>
        <dd>Plots the mutational signatures</dd>
      </dl><dl><dt>

          <code><a href="plot_umap.html">plot_umap()</a></code>

        </dt>
        <dd>Plot a UMAP from a musica result</dd>
      </dl><dl><dt>

          <code><a href="plot_cluster.html">plot_cluster()</a></code>

        </dt>
        <dd>Visualize clustering results</dd>
      </dl><dl><dt>

          <code><a href="plot_differential_analysis.html">plot_differential_analysis()</a></code>

        </dt>
        <dd>Compare exposures of annotated samples</dd>
      </dl><dl><dt>

          <code><a href="plot_heatmap.html">plot_heatmap()</a></code>

        </dt>
        <dd>Plot heatmaps using the exposures matrix</dd>
      </dl><dl><dt>

          <code><a href="plot_sample_counts.html">plot_sample_counts()</a></code>

        </dt>
        <dd>Plot distribution of sample counts</dd>
      </dl><dl><dt>

          <code><a href="plot_sample_reconstruction_error.html">plot_sample_reconstruction_error()</a></code>

        </dt>
        <dd>Plot reconstruction error for a sample</dd>
      </dl><dl><dt>

          <code><a href="predict_exposure.html">predict_exposure()</a></code>

        </dt>
        <dd>Prediction of exposures in new samples using pre-existing signatures</dd>
      </dl></div><div class="section level2">
      <h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a></h2>

      <div class="section-desc"><p>Functions to compare the signatures between two results objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="create_umap.html">create_umap()</a></code>

        </dt>
        <dd>Create a UMAP from a model result</dd>
      </dl><dl><dt>

          <code><a href="umap.html">umap()</a></code> <code><a href="umap.html">`umap&lt;-`()</a></code>

        </dt>
        <dd>Retrieve umap list from a result_model, result_collection, or musica object</dd>
      </dl><dl><dt>

          <code><a href="cluster_exposure.html">cluster_exposure()</a></code>

        </dt>
        <dd>Perform clustering analysis from a musica result object</dd>
      </dl><dl><dt>

          <code><a href="k_select.html">k_select()</a></code>

        </dt>
        <dd>Plots for helping decide number of clusters</dd>
      </dl></div><div class="section level2">
      <h2 id="signature-comparisons">Signature comparisons<a class="anchor" aria-label="anchor" href="#signature-comparisons"></a></h2>

      <div class="section-desc"><p>Functions to compare the signatures between two results objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="compare_cosmic_v2.html">compare_cosmic_v2()</a></code>

        </dt>
        <dd>Compare a result object to COSMIC V2 SBS Signatures (combination whole-exome and whole-genome)</dd>
      </dl><dl><dt>

          <code><a href="compare_cosmic_v3.html">compare_cosmic_v3()</a></code>

        </dt>
        <dd>Compare a result object to COSMIC V3 Signatures; Select exome or genome for SBS and only genome for DBS or Indel classes</dd>
      </dl><dl><dt>

          <code><a href="compare_results.html">compare_results()</a></code>

        </dt>
        <dd>Compare two result files to find similar signatures</dd>
      </dl><dl><dt>

          <code><a href="exposure_differential_analysis.html">exposure_differential_analysis()</a></code>

        </dt>
        <dd>Compare exposures of annotated samples</dd>
      </dl></div><div class="section level2">
      <h2 id="cosmic-signatures">COSMIC signatures<a class="anchor" aria-label="anchor" href="#cosmic-signatures"></a></h2>

      <div class="section-desc"><p>Result objects that contain COSMIC signatures</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="cosmic_v2_sigs.html">cosmic_v2_sigs</a></code>

        </dt>
        <dd>COSMIC v2 SBS96 Signatures Result Object</dd>
      </dl><dl><dt>

          <code><a href="cosmic_v3_dbs_sigs.html">cosmic_v3_dbs_sigs</a></code>

        </dt>
        <dd>COSMIC v3 DBS Genome Signatures Result Object</dd>
      </dl><dl><dt>

          <code><a href="cosmic_v3_indel_sigs.html">cosmic_v3_indel_sigs</a></code>

        </dt>
        <dd>COSMIC v3 Indel Genome Signatures Result Object</dd>
      </dl><dl><dt>

          <code><a href="cosmic_v3_sbs_sigs.html">cosmic_v3_sbs_sigs</a></code>

        </dt>
        <dd>COSMIC v3 SBS96 Genome Signatures Result Object</dd>
      </dl><dl><dt>

          <code><a href="cosmic_v3_sbs_sigs_exome.html">cosmic_v3_sbs_sigs_exome</a></code>

        </dt>
        <dd>COSMIC v3 SBS96 Exome Signatures Result Object</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous-functions">Miscellaneous functions<a class="anchor" aria-label="anchor" href="#miscellaneous-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="cosmic_v2_subtype_map.html">cosmic_v2_subtype_map()</a></code>

        </dt>
        <dd>Input a cancer subtype to return a list of related COSMIC signatures</dd>
      </dl></div><div class="section level2">
      <h2 id="data-objects-and-classes">Data objects and classes<a class="anchor" aria-label="anchor" href="#data-objects-and-classes"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="musica_sbs96.html">musica_sbs96</a></code>

        </dt>
        <dd>musica_sbs96</dd>
      </dl><dl><dt>

          <code><a href="musica_sbs96_tiny.html">musica_sbs96_tiny</a></code>

        </dt>
        <dd>musica_sbs96_tiny</dd>
      </dl><dl><dt>

          <code><a href="dbs_musica.html">dbs_musica</a></code>

        </dt>
        <dd>dbs_musica</dd>
      </dl><dl><dt>

          <code><a href="indel_musica.html">indel_musica</a></code>

        </dt>
        <dd>indel_musica</dd>
      </dl><dl><dt>

          <code><a href="musica.html">musica</a></code>

        </dt>
        <dd>musica</dd>
      </dl><dl><dt>

          <code><a href="musica_annot.html">musica_annot</a></code>

        </dt>
        <dd>musica_annot</dd>
      </dl><dl><dt>

          <code><a href="res.html">res</a></code>

        </dt>
        <dd>res</dd>
      </dl><dl><dt>

          <code><a href="res_annot.html">res_annot</a></code>

        </dt>
        <dd>res_annot</dd>
      </dl><dl><dt>

          <code><a href="rep_range.html">rep_range</a></code>

        </dt>
        <dd>Replication Timing Data as GRanges Object</dd>
      </dl><dl><dt>

          <code><a href="musica-class.html">musica-class</a></code>

        </dt>
        <dd>The primary object that contains variants, count_tables, and samples annotations</dd>
      </dl><dl><dt>

          <code><a href="count_table-class.html">count_table-class</a></code>

        </dt>
        <dd>Object containing the count table matrices, their names and descriptions that we generated by provided and by user functions. These are used to discover and infer signatures and exposures.</dd>
      </dl><dl><dt>

          <code><a href="result_collection-class.html">result_collection-class</a></code>

        </dt>
        <dd>The Result Collection object that contains modality, input parameters, prior hyperparameters</dd>
      </dl><dl><dt>

          <code><a href="result_model-class.html">result_model-class</a></code>

        </dt>
        <dd>Object that contains results for a single model</dd>
      </dl><dl><dt>

          <code><a href="built_tables.html">built_tables()</a></code>

        </dt>
        <dd>Retrieve the names of count_tables from a musica object</dd>
      </dl><dl><dt>

          <code><a href="musicatk.html">musicatk()</a></code>

        </dt>
        <dd>Starts the musicatk interactive Shiny app</dd>
      </dl><dl><dt>

          <code><a href="rc.html">rc()</a></code>

        </dt>
        <dd>Reverse complement of a string using biostrings</dd>
      </dl><dl><dt>

          <code><a href="table_selected.html">table_selected()</a></code>

        </dt>
        <dd>Retrieve table name used for plotting from a result_model object</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aaron Chevalier, Natasha Gurevich, Tao Guo, Joshua D. Campbell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

